architecture_records/004-datapreprocessor.md

```diff
@@ -0,0 +1,421 @@
+
+1. Keep intermediate datasets in memory while preprocessing using [`keep_in_memory=True`](https://huggingface.co/docs/datasets/v3.1.0/en/package_reference/main_classes#datasets.Dataset.map.keep_in_memory), for large datasets and Iterable datasets we assume this to be for mini batches.
+1. Disable caching in file system and make it configurable by [`disable_caching()`](https://huggingface.co/docs/datasets/en/cache#enable-or-disable-caching) API from HuggingFace.
+
+These considerations can be made dynamicallly or can be passed via users as we allow any `kwargs` to be passed to the map operations.
+
+## Alternatives Considered
+
+### Letting users process their own data and pass file(s) directly to this library.
+
+A simple alternative to avoid all this is to have the users process their own data, this is also in lines of the fact that most workloads contain preprocessed data which is used by simple users as is for their tuning/training.
+
+The reason to have this design is that many users coming to this library have advanced set of use cases. As stated in the motivation we are getting ever increased demand from researchers looking to use this library are looking for features like `jinja template` rendering, image data processing, mixing and merging datasets. While this can be done at user level most users are not looking to write code to do all this preprocessing but use tools which implement them to perform these tasks. 
+Leaving all users to write their own preprocessing logic can also lead to code duplication across many teams which is something we want to avoid.
+
+More importantly as stated in the motivation we are getting ever increased demand from users who want to use this library directly with their dataset and have quick roundtrip for testing. This design allows users to specify simple parameters in the config and test for complex usecases easily.
+
+### Passing all datasets we take to the Huggingface SFTTrainer api and let it handle them without preprocessing at our end.
+
+Another alternative we have is to take the `dataset` input to this library and pass it directly to the trainer `SFTrainer` in our case directly and let it handle loading and preprocessing the dataset.
+
+[SFTrainer](https://huggingface.co/docs/trl/v0.12.1/en/sft_trainer#trl.SFTTrainer) supports specifying the `train_dataset` and `eval_dataset` for both of which it supports iterable datasets along with normal datasets allowing us to pass a large dataset supported via streaming. 
+
+Please not that even in this case users will need to tell us that the dataset is large and is to be loaded via `streaming=True` because the argument which tells HF to load the dataset in iterable mode or standard mode is passed to [`load_dataset`](https://huggingface.co/docs/datasets/v3.1.0/en/package_reference/loading_methods#datasets.load_dataset)
```

Comment:
nit:
```suggestion
Please note that even in this case users will need to tell us that the dataset is large and is to be loaded via `streaming=True` because the argument which tells HF to load the dataset in iterable mode or standard mode is passed to [`load_dataset`](https://huggingface.co/docs/datasets/v3.1.0/en/package_reference/loading_methods#datasets.load_dataset)
```